[Blackjack - Step2] Submit miooo0o's mission

miooo0o opened the pull request:
## Additional Updated Features
### Betting System Implementation:
- Add PlayerFactory to handle player creation with betting amounts
- Implement input validation for player names and bet amounts
- Add Validator class to ensure proper name and amount validation
- Support reading player bet amounts through InputView

### Payout and Result System:
- Add PayoutMultiplier property to Outcome enum for different win conditions
- Implement PlayerResult class with finalAmount calculation for individual player payouts
- Implement DealerEarning class to track dealer's total earnings
- Integrate PlayerResult and DealerEarning into comprehensive GameResult system

### Game Logic Improvements:
- Fix Blackjack detection logic to properly identify 21 with initial two cards
- Add dealer hit/stand logic with proper state transitions
- Implement edge case handling for various game outcomes
- Separate val/var usage to prevent state mutation issues

### Testing and Code Quality:
- Move test-specific classes (DummyDeck, DummyPlayerFactory) to test package
- Improve naming and readability 
- Collect view logic from Card and Suit classes into OutputView
- Reduce method visibility to internal where appropriate
- Add comprehensive test coverage for gameplay scenarios and edge cases

---

## Reflection

This is the first commit of step2, and I learned a lot about why forcing inheritance like this is problematic. What you mentioned about those 5 issues turned into an endless cycle when I tried to fix just the errors discovered in the process.

The most terrible part was the `state (var/val)` configuration being tangled up due to the `receive` method. Fixing one thing would break another, so I spent the entire day writing tests, writing more tests, writing even more tests, fixing, and repeating. Honestly, I'm still very anxious about it. Even though all tests pass, I'm worried.

Before moving to step2, I had prepared a revert version that only fixed what was reviewed. I wonder if I should have said I'd go with that version instead, given how painful this was. However, thanks to this struggle, I'm now more determined to avoid making the structure complicated!

### What I Regret
- It took so long to fix errors and the complex relationships made it hard to catch up once I lost focus, so I couldn't update the README in time, which is disappointing.
- I thought it would be better to separate `val playerResults: List<PlayerResult> = getAllPlayerResults()` into its own class (since there's evaluation logic, it seemed better to extract it), but due to time constraints, I decided to send the PR first. I'm curious about your thoughts on this.

### What I'd Like Feedback On
- I tried to make it at least a bit more readable without changing the structure too much. Do you think it improved a little?
- I did untangle the `state (var/val)` and `receive` (previously, receive directly mutated state), but I felt there were too many places tracking state changes with `get()`. Even when running the debugger, it was hard for me to follow... For values that "need to continuously change or be updated," what would be a good way to organize them?
- I'm sure there are probably more errors somewhere. During this change, I barely thought about questions like "why is this class a class?" I was just grateful when things worked and was afraid to change more... I learned again that not writing tests properly leads to pain. If you notice any areas I should cover better, please let me know.

VittorioDeMarzi committed:
feat(domain): implement operator plus in cards class

VittorioDeMarzi committed:
test(domain): add test for cardDeck shuffling

VittorioDeMarzi committed:
feat(domain): implement shuffle logic for cardDesk

VittorioDeMarzi committed:
refactor(domain): rename the class Cards to Hold

VittorioDeMarzi committed:
refactor(domain): change visibility of property to private and change name Cards class to Hold

VittorioDeMarzi committed:
test(domain): refactor name of test

VittorioDeMarzi committed:
test(domain): refactor test since priority of holds are private

VittorioDeMarzi committed:
refactor(domain): add methods in CardDeck to retrieve number of cards and list of cards

VittorioDeMarzi committed:
feat(domain): implement Hand class

VittorioDeMarzi committed:
test: refactor tests to moveCard to a Player

VittorioDeMarzi committed:
feat: implement Player class

VittorioDeMarzi committed:
feat: implement Person wrapper class for names

VittorioDeMarzi committed:
refactor(Hold): change target of moveCard to Player

VittorioDeMarzi committed:
docs(readme): update the progress

VittorioDeMarzi committed:
refactor: refactor packages to controller/model/view

VittorioDeMarzi committed:
docs(readme): add questions

VittorioDeMarzi committed:
feat(domain): add game controller and basic logic

VittorioDeMarzi committed:
test(domain): add new test for player factory class

VittorioDeMarzi committed:
feat(domain): implement PlayerFactory object

VittorioDeMarzi committed:
refactor(domain): add constructor to Player

VittorioDeMarzi committed:
test: refactor test to encapsulate the value

VittorioDeMarzi committed:
refactor: change visibility to property

VittorioDeMarzi committed:
refactor: implement method to create a dealer

VittorioDeMarzi committed:
refactor: increase readability of class Hand

VittorioDeMarzi committed:
test: add test to hit a card from the desk to a player

VittorioDeMarzi committed:
feat: implement hit method

VittorioDeMarzi committed:
refactor: remove unnecessary constructor and merge

VittorioDeMarzi committed:
refactor: adopt previews refactor

VittorioDeMarzi committed:
feat: implement GameManager class

VittorioDeMarzi committed:
test: add a new test to set up the table with GameManager

VittorioDeMarzi committed:
feat: implement GameManager and init logic

- dealer has a card
- each player have 2 cards

VittorioDeMarzi committed:
test(Hand): add tests for counting the points in one hand of a player

VittorioDeMarzi committed:
feat: implement logic for counting points in the hand of a player

VittorioDeMarzi committed:
feat: implement data class for a single player

VittorioDeMarzi committed:
refactor: adopt based to the previews feat

VittorioDeMarzi committed:
test: add test for get single player result

VittorioDeMarzi committed:
feat: get single player result and play with multiple players

VittorioDeMarzi committed:
refactor: adopt core logic to controller

VittorioDeMarzi committed:
refactor: add Rank to card

VittorioDeMarzi committed:
refactor: adapt to previews refactor

VittorioDeMarzi committed:
feat: add input validation for players names

VittorioDeMarzi committed:
feat: add Errors enum messages

VittorioDeMarzi committed:
refactor:remove card test not needed

VittorioDeMarzi committed:
refactor: adapt controller to logic of InputView

VittorioDeMarzi committed:
refactor: override toString

VittorioDeMarzi committed:
refactor: override toString and refactor get()

VittorioDeMarzi committed:
refactor: delete playerResult class

VittorioDeMarzi committed:
refactor: adapt controller and manager to previews refactor

VittorioDeMarzi committed:
feat: Output of first SetUp

VittorioDeMarzi committed:
refactor: ask for card In Input View

VittorioDeMarzi committed:
refactor: change logic of the game separating play with player or with dealer

VittorioDeMarzi committed:
refactor: Output View

VittorioDeMarzi committed:
refactor: adapt the controller to the previews refactor

VittorioDeMarzi committed:
refactor: add to player 2 properties, isBlackJack and isBust

VittorioDeMarzi committed:
feat: add and implement logic of statistics class

VittorioDeMarzi committed:
feat: print final result statistics

VittorioDeMarzi committed:
refactor: remove toString override

VittorioDeMarzi committed:
refactor: protect helper methods to private or internal: if needed to test

VittorioDeMarzi committed:
refactor: adopt prev commit

- change name of method inside GameManager

VittorioDeMarzi committed:
fix: remove not related class

VittorioDeMarzi committed:
refactor: add more logic to check user input

- check a result of split input as well

VittorioDeMarzi committed:
style: remove comments

VittorioDeMarzi committed:
refactor: adopt Statistics class

VittorioDeMarzi committed:
docs: update the progress

VittorioDeMarzi committed:
refactor: remove unused variable and constructor

miooo0o committed:
refactor: change GameManager.kt package directory

miooo0o committed:
feat: plus operator in Player.kt

miooo0o committed:
refactor: adopt prev feature

miooo0o committed:
refactor: change PlayerFactory.with() to createPlayers()

- if apply interface -> Player/Dealer logic, I would like to change method name to with() again, and remove createDealer methods

miooo0o committed:
refactor: change to emptySet()

miooo0o committed:
refactor: separate model to sub-dirs

miooo0o committed:
feat: create new Player, change Old-Player class name to PlayerBackup

miooo0o committed:
test: create new Test for Participant/Player/Dealer, and Participants

miooo0o committed:
feat(model): implement new abstract class Participant

miooo0o committed:
feat(model): implement Participants class

- has Dealer
- with name or list of players, create "Participants",

miooo0o committed:
test: apply new class to ParticipantTest.kt

miooo0o committed:
refactor: rename Old-Player parameter

miooo0o committed:
test: add new test to check containAll method

miooo0o committed:
feat: implement containAll method for test in Participants

miooo0o committed:
test: implement containAll method for test in Participants

- will use `contain(Participant)`
- other `contain` logic methods should be separated into test functions, if possible.

miooo0o committed:
feat: implement Participants class

- class initiated with two constructors: with name of players or list of players
- should track class constructor usage

miooo0o committed:
test: refactor test to way to init `Dealer` for module test

miooo0o committed:
feat: encapsulate dealer instance

miooo0o committed:
test: add new test for recognize number of players in Participants

miooo0o committed:
feat: implement method to check the number of players

miooo0o committed:
refactor: change directories for refactor

miooo0o committed:
test: add new test to fit with controller parameter in Participant

such as
- Participants has two constructors which have `String/Player` as parameter, but InputView returning `List<String>`
- change directories for refactor

miooo0o committed:
feat: add companion object method to avoid JVM constructor signature clash

miooo0o committed:
test: refactor methods name, way to initiate in Participants test

- `Participants` has private constructor
- `Participants` only initiated `from`

miooo0o committed:
refactor: way to initiate/methods in Participants

- `Participants` has private constructor
- `Participants` only initiated `from`
- `players` and `dealer` are private, and getters have same naming logic

miooo0o committed:
feat: implement a new Controller

- BlackJackGame will replace BlackJackController

miooo0o committed:
refactor: change method name inside BlackJackGame, `play()` to `run()`

miooo0o committed:
refactor: remove all methods and remain base elements

miooo0o committed:
refactor: apply new method name run() in Application

miooo0o committed:
feat: new print error method in OutputView

miooo0o committed:
feat: implement CardHolder class for Hand/CardDeck

miooo0o committed:
feat: add new method size() in CardHolder

miooo0o committed:
refactor: remove useless files at current process

miooo0o committed:
test: remove not usable test

miooo0o committed:
test: separate fixture object to TestFixture.kt

miooo0o committed:
refactor: Apply the change in the package of Card Deck to GameManager

miooo0o committed:
test: add new test for an edge case that Hand try to `draw()` when List is empty

miooo0o committed:
feat: implement Hand, Child class that inherits CardHolder

- if drawCard fail, onDrawFailed() called as common logic, and throw exception

miooo0o committed:
docs: create a new list for step 1 refactor

miooo0o committed:
docs: update the list for step 1 refactor

miooo0o committed:
refactor(model): change methods name to more readable

- drawCard() -> draw(): Card
- receiveCard() -> receive(Card)

miooo0o committed:
test: refactor TestFixture class to object

miooo0o committed:
test: add new tests for CardHolder core logics.

miooo0o committed:
feat: add new method `peek()` to CardHolder, which peek first card in List<Card>

miooo0o committed:
test: create a new test to Deck

miooo0o committed:
feat: implement Deck class.

- if Deck empty, Deck generate new cards, List<Card>
- init with Shuffled deck
- refactor 1st version of Deck. property `cards` is List from Set for generate new cards

miooo0o committed:
refactor: CardHolder property `cards` is variable for generate new card logic in Deck

miooo0o committed:
test: add a test for Deck. check Deck is created by initDeck() method

miooo0o committed:
refactor: apply new Deck in GameManager

miooo0o committed:
style: add EOF

miooo0o committed:
test: fix test which has a wrong index in ParticipantTest

- init with players()
- participants.getPlayer(1).name -> participants.getPlayer(0).name

miooo0o committed:
test: refactor test to check deck generate a new deck

- Previously test used the values of List<Card>, but now check the reference

miooo0o committed:
refactor: move println() from Deck to Output so that OutputView can have more accurate responsibility.

miooo0o committed:
test: move TestFixture object to model

miooo0o committed:
feat: add getter in Participants

miooo0o committed:
test: change class ParticipantTest to ParticipantsTest

miooo0o committed:
test: create ParticipantTest to test common logic between Participant subclass

miooo0o committed:
refactor: Participant inherits from Hand

miooo0o committed:
refactor: Divide peek() into first() and last() for separate usage in Hand

miooo0o committed:
test: add a new test for Player

- Player should have a correct name format

miooo0o committed:
refactor: remove unused method

miooo0o committed:
feat: implement player name rule

- should letter

miooo0o committed:
refactor(view): collect println() to OutputView

miooo0o committed:
feat: implement retry logic

miooo0o committed:
test: change location of TestFixture to blackjack

miooo0o committed:
test: change location of TestFixture to blackjack

miooo0o committed:
test: add new test for InputView

miooo0o committed:
refactor: change the input view to a testable structure

miooo0o committed:
feat: add new message string in OutputView

miooo0o committed:
refactor: to make it easier to apply Retryable, split the InputView into `ask` and `parser`, and then wrap them in `read` to organize them

miooo0o committed:
test: add a new test for InputView, Parsing process

miooo0o committed:
refactor: remove duplicated loing in Player, and give a basic validation process to InputView-Parser

- Parser should be separated later
- Ask as well (consider)

miooo0o committed:
refactor: took the logic used in the first PR and refactored it

miooo0o committed:
refactor: took the logic used in the first PR and refactored it to Deck

- hit: draw first of Deck.cards, and Participant receives it

miooo0o committed:
revert: get lines from step 1 first PR in GameManager

miooo0o committed:
test: change method name in CardHolder

miooo0o committed:
refactor: apply new data structures to GameManager

miooo0o committed:
refactor: apply new data structures to OutputView

miooo0o committed:
feat: add new variable state

- was included with the intention of a future refactor, but there is no guarantee that new logic will adopt (limited time)

miooo0o committed:
style(state)

miooo0o committed:
refactor: delete a test in Player due to logic removed

miooo0o committed:
feat(player): implement evaluateState logic in Player

- if player card points is 21 in the first round, blackjack
- if over, bust
- else -> controlled my GameManager with user input

miooo0o committed:
feat(dealer): implement evaluateState logic

- Return BLACKJACK when dealer has 21 points initially
- Return BUST when points exceed 21
- Return HIT when points < 17 (dealer must draw)
- Return STAY when points >= 17

miooo0o committed:
refactor: const vals that are used repeatedly are bundled into an object, GameConstants.kt

miooo0o committed:
refactor: changed methods names in Output View to make them easier to read and applied to related classes

miooo0o committed:
refactor: add getter in Participant, state

miooo0o committed:
refactor: renamed the methods to make it more readable.

miooo0o committed:
refactor: changed GameManager to be able to get the final result after running the game.

miooo0o committed:
refactor: apply prev logic, and change show method name

miooo0o committed:
fix: an issue state in Player is going into an infinite loop when state is changed.

miooo0o committed:
test: add a new test for GameManager

miooo0o committed:
fix: an issue hit() In Deck didn't repeat as much

miooo0o committed:
feat: apply how to display the GameResult to be used in the future

- still need more unit tests, couldn't follow TDD

miooo0o committed:
fix: method name Participant in compiler recognizes get() as duplicate

miooo0o committed:
feat: implement class that creates GameReuslt

miooo0o committed:
feat: implement enum class Outcome to rank final result and show/print prefix

miooo0o committed:
feat: Add a method that prints the final results

miooo0o committed:
docs: update the progress

miooo0o committed:
refactor: remove unused parameter

miooo0o committed:
refactor: comment-out

miooo0o committed:
fix: logic for Dealer to decide Stay/Hit based on the number of cards, not the point

miooo0o committed:
refactor: rename property `points` to `score` to improve readability.

miooo0o committed:
style: apply formatting

miooo0o committed:
style: apply formatting

miooo0o committed:
refactor: remove unused class, Error

miooo0o committed:
test: add test for GameManager

miooo0o committed:
test: add new test helper object, Card

miooo0o committed:
refactor: Remove plural suffix from Suit enum constants

- HEARTS → HEART
- DIAMONDS → DIAMOND
- CLUBS → CLUB
- SPADES → SPADE

miooo0o committed:
test: apply renamed Suit enum constants

miooo0o committed:
test: apply renamed Suit enum constants

miooo0o committed:
refactor: to make a bit safer for Players and Dealers to receive, we added conditions and modified `receive` return type to Boolean

miooo0o committed:
test: add new test for round works correctly in GameManager

miooo0o committed:
fix: update logic to eval state with score itself when Participant.receive() call

miooo0o committed:
refactor: rename cardDeck to deck in GameManager, and add new internal properties for test

miooo0o committed:
refactor: delete the empty test file, PlayerTest.kt

miooo0o committed:
test: refactor InputView.Parser to Parser

miooo0o committed:
test: move methods which only used in Test -- "first(), last()" to TestFixture

miooo0o committed:
refactor: rename GameConstants name

- ABLE_TO_RECEIVE to DEALER_MUST_HIT_LIMIT

miooo0o committed:
refactor: remove first(), last() methods in InputView

miooo0o committed:
refactor: apply renamed const name

miooo0o committed:
docs: init a new feature list for step2

miooo0o committed:
refactor: change draw/receive logic in CardHolder

- draw return List<Card>
- receive manage List<Card>

miooo0o committed:
refactor: remove override onDrawFailed() in Hand/Deck

- onDrawFailed() merged into CardHolder, and if "cards" are empty, throw IllegalStateException

miooo0o committed:
refactor: remove unused print method in OutputView

miooo0o committed:
test: apply new logic from prev refactoring

- empty deck throw illegal exception
- draw/receive multiple cards

miooo0o committed:
style: apply formatting

miooo0o committed:
docs: update the progress

miooo0o committed:
refactor(participants): make properties public readonly

- Replace private val + getter pattern with public val
- Simplifies property access for players and dealer

miooo0o committed:
refactor(application): remove unused controller and move logic to BlackJackGame.

miooo0o committed:
test(test-fixture): clean up

miooo0o committed:
refactor(participant): add new parameter amount, and apply to related methods

miooo0o committed:
chore(participants): add note

miooo0o committed:
test(manager, participants): ready for refactor

- improve test, tests use "FakePlayerFactory"

miooo0o committed:
refactor (manager, game-result, view): replace getter players/dealer to properties itself

miooo0o committed:
feat(player-factory): implement PlayerFactory

- BasePlayerFactory: interface for player creation
- PlayerFactory: creates list of Player instances
- FakePlayerFactory: generates dummy players for testing (configurable count)

miooo0o committed:
test(refactor): apply renamed DummyPlayerFactory

miooo0o committed:
refactor(dealer): rename const value to DEALER_NOT_BET_MONEY

miooo0o committed:
feat(input-view): implement read methods to get player's bet amount

- amount should be positive number

miooo0o committed:
refactor(application): apply PlayerFactory to application

miooo0o committed:
test(validator): add tests to name validation

miooo0o committed:
feat(validator): implement name, amount validation

miooo0o committed:
test(validator): add new tests to amount validation

miooo0o committed:
feat(input-view): apply new feature Validator

miooo0o committed:
refactor(output-view): update show methods

- to clean up code, separate parameters or rename

miooo0o committed:
refactor(factory): separate DummyPlayerFactory to test package

miooo0o committed:
refactor(deck): separate DummyDeck to test package

miooo0o committed:
refactor: rename Constants value to improve readability

miooo0o committed:
chore: clean up

miooo0o committed:
test: add a new test for confirmation player state

miooo0o committed:
test(deck): add a new test to check shuffledDeck

miooo0o committed:
feat(outcome): add payoutMultiplier property to Outcome enum

- payoutMultiplier calculates total payout (base + winnings)
- represents rate as (current amount + additional winnings)

miooo0o committed:
test: apply DummyPlayerFactory

miooo0o committed:
chore(participants): reduce method visibility to internal

- tests only access methods

miooo0o committed:
test(player-result, dealer-earning): add comprehensive payout tests

- add PlayerResult tests for all outcome scenarios
- add DealerEarning tests for corresponding dealer profits/losses
- verify the correct calculation for WIN, LOSE, DRAW, BLACKJACK outcomes

miooo0o committed:
feat(player-result): implement PlayerResult with finalAmount calculation

- calculates final payout based on bet amount and outcome
- supports WIN, LOSE, DRAW, and BLACKJACK scenarios

miooo0o committed:
feat(dealer): implement DealerEarning with dealer amount calculation

- calculates final payout based on player's final amount

miooo0o committed:
chore: apply renamed const value, FIRST_ROUND_HIT_COUNTS

miooo0o committed:
chore: clean up application

miooo0o committed:
refactor(game-result): integrate PlayerResult and DealerEarning into GameResult

miooo0o committed:
chore(dealer-earning): remove unused variable

miooo0o committed:
refactor(validator): separate validate logics from PlayerFactory

miooo0o committed:
refactor(participant): improve readability

- Participant subclasses have boolean methods, "isBust()" "isBlackjack()"
- Remove Participant init validation, all names/amounts will be provided by Validator before being passed as parameters
- remove unnecessary state value/getter()

miooo0o committed:
refactor(GameResult): improve code clarity with participant boolean methods

- use isBust() and isBlackjack() methods from participant classes to make the code cleaner and create more precise branching logic

miooo0o committed:
fix(GameResult): add a condition that check player's blackjack before state branch moved to win.

miooo0o committed:
test(game-result): add new tests to cover gameplay scenarios

miooo0o committed:
test(test-fixture): rename card deck to improve readability

miooo0o committed:
fix(game-result): fix condition to correctly assign Blackjack outcome

miooo0o committed:
docs(readme): clean up document

miooo0o committed:
refactor(output-view, card, suit): collect view logic from card, suit

- implement extension function to card/suit

miooo0o committed:
docs(readme): update the progress

miooo0o committed:
test(game-result): add tests to confirm dealer hit/stand logic

miooo0o committed:
test(dealer, game-manager): fix tests where dealer incorrectly keeps drawing after the score exceeds 16

miooo0o committed:
fix(participant, player, dealer): separate val/var usage to prevent state mutation in receive()

- fixed a critical design bug where `state` was dynamically computed, causing unpredictable behavior during `receive()`.
- separated state into mutable `mutableState` and val-accessed `state` to ensure consistency and correctness.

miooo0o committed:
fix(game-manager): prevent player/dealer from drawing when deck is empty

miooo0o committed:
chore: add new cards prefix

miooo0o committed:
test(game-result): add test cases for edge outcomes including blackjack, bust, and draw

miooo0o committed:
docs(readme): update the progress

miooo0o committed:
refactor(constants): move DEALER_NOT_BET_MONEY from Dealer

miooo0o committed:
docs(readme): improve readability

miooo0o committed:
Merge remote-tracking branch 'tech/miooo0o' into step2

pkeugine commented on src/main/kotlin/blackjack/factory/PlayerFactory.kt:
```diff
@@ -0,0 +1,20 @@
+
+interface BasePlayerFactory {
+    fun createPlayers(): List<Player>
+}
```
```suggestion
fun interface BasePlayerFactory {
    fun createPlayers(): List<Player>
}
```
We can make this a functional interface if we have only one method in it.
It would be great to check out what functional interface is :)

miooo0o replied:
As I understand it so far, a `fun interface` allows us to pass a lambda.
I’m not exactly sure how I would adapt that in the current version. 
But if I ever need a new kind of test, I’ll be able to experiment with a lambda instead of creating a new class. right?

For now, I’ve just converted the `interface` to a `fun interface`.
Please have a look, and let me know if you’d recommend writing a test to try it out!

pkeugine replied:
Yes, you can use it in lambda 👍
That's it! You don't need to do anything.

pkeugine commented on src/main/kotlin/blackjack/model/holder/CardHolder.kt:
```diff
@@ -9,27 +9,26 @@ abstract class CardHolder {
+        return cards
+    }
+
+    /** fallback when cannot draw */
```
Since `onDrawFailed` seems like it describes itself pretty well, I don't think we need this comment here. Let's keep that in mind when working on other projects.

You might have noticed I put very close attention to comments in between logics.
Although it might seem like they're not that big of a deal, I prefer adding comments when I feel they are **extremely necessary**. If not, the whole team starts to add in comments and they become part of maintaining. That leads to deprecated comments, or comments that are only understandable to certain people. This could be a real pain, because it's not always clear if I can remove that comment or not, resulting in confusing comments staying there forever.

I do understand that `fallback when cannot draw` seems like it won't confuse us in the future. But I thought I would explain why I pay close attention to comments since I've left reviews about it in the previous PR as well.

miooo0o replied:
I see that proper documentation really makes a difference. I used to just write documents from time to time, but intentional documentation is definitely something else! 
As always, keep it simple and understandable through code only.

Thanks for the advice.

pkeugine commented on src/main/kotlin/blackjack/model/participant/Dealer.kt:
```diff
@@ -1,33 +1,42 @@
+    internal fun calculateState(): State {
+        if (isFirstRound()) return initialStateOrBlackjack()
+        return evaluateState()
     }
```
I've noticed you've been using internal access level for some methods.
I'm guessing this is due to test availability for private methods.

However, some of these internal methods aren't tested, and are internal but not used in anywhere else.
I would keep them private, and focus on testing public methods that use these private methods.
If we make good tests and cover the edge cases for public method tests, I'm sure we'll be able to cover private methods as well.

miooo0o replied:
I tried my best to find and replace as many as I could!

pkeugine commented on src/main/kotlin/blackjack/model/holder/CardHolder.kt:
```diff
@@ -9,27 +9,26 @@ abstract class CardHolder {
-            currentCards.removeFirst()
-        } else {
+    fun draw(count: Int): List<Card> {
+        if (currentCards.isEmpty() || currentCards.size < count) {
```
Nice!! 💯 💯 💯

pkeugine commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -1,62 +1,78 @@
-        playersResults.forEach { (player, outcome) ->
-            println("${player.name}: ${outcome.name}")
+    internal fun showGameSummary(gameResult: GameResult) {
+        println("\n${Message.FINAL_RESULTS_TITLE}")
+        println("Dealer: ${gameResult.dealerEarning.amount}")
```
We can use Kotlin's [multi-line string feature](https://kotlinlang.org/docs/java-to-kotlin-idioms-strings.html#use-multiline-strings) when printing multiple lines!
`println` is actually quite expensive when it comes to using computation resources, so we might as well decrease the amount of times calling it. It might be fun to search and learn why :)

miooo0o replied:
Oh wow, I never thought `println()` could be expensive. that’s really surprising! I definitely want to look into why that is. It’s super interesting, and I’d love to try applying it.

pkeugine commented on src/test/kotlin/blackjack/controller/GameManagerTest.kt:
```diff
@@ -15,29 +17,29 @@ import org.junit.jupiter.api.Test
-        val participants = Participants.from(listOf("mina", "guri"))
+        val participants = Participants(DummyPlayerFactory(2).players)
         GameManager(participants).setUp()
 
-        OutputView.showCards(participants.getDealer())
-        OutputView.showAllPayersCards(participants)
+        OutputView.showCards(participants.dealer)
+        OutputView.showAllPayersCards(participants.players)
```
Although I didn't leave this review in the previous PR, it would be great to have at least one assertion in tests instead of having UI result. This is because we don't actually see the whole test result logs whenever we run the tests. That means even if the output is wrong, the test will still pass and we won't notice it. In this case, I don't think this test is necessary, with reasons I've written in the previous comment.

miooo0o replied:
That's right. I completely forgot this test during refactor.
I'll have to spend some more time cleaning up the code. Thank you, I'll apply it!

pkeugine commented on src/test/kotlin/blackjack/model/participant/DealerTest.kt:
```diff
@@ -0,0 +1,55 @@
+        dealer.receive(TWO_CARDS_SUM_16)
+        dealer.receive(listOf(TestFixture.Card.DIAMOND_JACK))
+
+        println("cards: ${dealer.cards}")
+        println("score: ${dealer.score}")
+        println("state: ${dealer.state}")
```
Tests don't need println in their final form. They can be used when we try to debug a test, but let's keep them out of our main branch.

miooo0o replied:
Oh, this is the part I forgot to delete. I'll delete!

pkeugine commented on src/test/kotlin/blackjack/model/result/PlayerResultTest.kt:
```diff
@@ -0,0 +1,39 @@
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+class PlayerResultTest {
+    @Test
+    fun `final amount is -10000 if player bet 10000 and Lose`() {
+        val player = DummyPlayerFactory(1).players[0]
+        val playerResult = PlayerResult(player, Outcome.LOSE)
+
+        assertThat(playerResult.finalAmount).isEqualTo(-10000)
+    }
+
+    @Test
+    fun `final amount is 20000 if player bet 10000 and Win`() {
+        val player = DummyPlayerFactory(1).players[0]
+        val playerResult = PlayerResult(player, Outcome.WIN)
+
+        assertThat(playerResult.finalAmount).isEqualTo(20000)
+    }
```
I can see that you've put so much effort in testing 👍 

**PR message: Honestly, I'm still very anxious about it. Even though all tests pass, I'm worried.**
Don't be. You won't last so long as a developer with that anxiety in you. Just let it go, and fix when somebody finds a bug 😄

miooo0o replied:
I wanted to make sure I had at least the basic tests covered before PR.

I realized during the Lotto mission that if I write loose tests, I can go on for a long time thinking everything’s fine when it’s actually not. This time I was writing and editing tests again and I actually found a lot of tests that made no sense haha. but... sure, if something breaks, I can just fix it!

pkeugine commented on README.md:
```diff
@@ -2,66 +2,25 @@
-- The dealer must draw a card if their total is 16 or less, and must stand on 17 or more.
-- If the dealer busts (goes over 21), all remaining players automatically win.
-- After the game ends, display the result (win/loss) for each player.
+- [x] Each player must place a bet at the start of the game.
```
**PR message : thanks to this struggle, I'm now more determined to avoid making the structure complicated!**

**PR message 2 : I tried to make it at least a bit more readable without changing the structure too much. Do you think it improved a little?**

Yes, the project looks much better than before, and you've done a great job refactoring it. I wouldn't mind making it more simple though, like I said in the previous comments. No worries about testing private functions, no UI related testing when making unit tests, no comments when the code describes itself pretty well, etc. Personally, I like to think as : `Simple is better most of the times. Add in stuff when people say they need it` 😄

miooo0o replied:
In this set of commits, I removed all the test output that was previously being printed.It definitely looks much cleaner now. Yes, simple is better most of the time. happier, simpler and cleaner!

pkeugine commented on README.md:
```diff
@@ -2,66 +2,25 @@
 
 ## Requirements
```
**PR message : I thought it would be better to separate `val playerResults: List<PlayerResult> = getAllPlayerResults()` into its own class (since there's evaluation logic, it seemed better to extract it), but due to time constraints, I decided to send the PR first. I'm curious about your thoughts on this.**

Yes, I do get what you mean when you say there could be a class dedicated to `List<PlayerResult>`, but the current logic seems fine to me. I'm fine with keeping it as is. Though I think `DealerEarning` is the class that you're looking for; it just has a name related to `Dealer`. 

(I'm not saying it's wrong, so don't stress too much about refactoring this)

miooo0o replied:
Ah, I’m glad to hear that!  I’ll have to extract the dealer logic separately. I’ll go ahead and do that.

Coding itself doesn’t stress me out! If anything, all the overthinking and constant questioning is just a leftover occupational habit from the past. Honestly, I kind of enjoy that kind of stress :sparkles: 

I wanted to hit my goals (like sticking to the basics and improving test quality), so I was a bit worried that I didn’t quite get there.  But aside from that, I genuinely enjoyed it.

pkeugine commented on README.md:
```diff
@@ -2,66 +2,25 @@
 
 ## Requirements
```
**PR message : For values that "need to continuously change or be updated," what would be a good way to organize them?**

For values that need to change over time, using var is typically appropriate 😅.
Alternatively, if you want to keep things immutable, consider returning a new object with the updated values instead of mutating the existing one, a common pattern in functional programming.

Most of the times in web backend development, we use our database 😎

miooo0o replied:
This is my first time doing something where there is a distinction between `var` and `val`, so I felt like I shouldn't use `var`. I don't know why but :crying_cat_face: 

If you use a database in the backend, there must be a convention, right? You mean I have less to worry about? It would be great.

pkeugine replied:
I think we'll cover this as we progress through the next mission :)

pkeugine commented on README.md:
```diff
@@ -2,66 +2,25 @@
 
 ## Requirements
```
**PR message : During this change, I barely thought about questions like "why is this class a class?" I was just grateful when things worked and was afraid to change more... I learned again that not writing tests properly leads to pain. If you notice any areas I should cover better, please let me know.**

I'm continuously noticing you're thinking a bit too deep about this.
Let's try to keep things (including logic) simple, not worry too much about it, and let the problem occur... cuz we're never free from bugs 🙂

Hey, even the inventors of C language, Linux, and all these popular software created bugs and legacy codes that others couldn't deal with. This happens to everyone, and nobody is free from it. So we might as well be cool with it.

miooo0o replied:
It kind of feels like if I don't deal with it now, it'll come back to haunt me later like karma. still but slowly I let go.
This time, not overthinking actually made things a bit easier. Just letting it be like, “Okay, this works, note and move on” felt really good.

I guess learning to let go of the pressure is part of growing :confetti_ball:

pkeugine reviewed (changes requested):
Hello Mina,
Well done keeping this mission going even when you're busy with the next mission.
I tried to keep the review simple, since it is a fact that `we're currently working on the next mission`. However, I thought you might want to look at the comments and refactor a bit, since I know you well (... I think).

But I want to give you an additional goal:
**Keep it under 5 commits**.

Once you're done, give me an additional review request, and I'll merge this PR 👍 

Happy Hacking, and remember, **keep it simple**. 😎

miooo0o committed:
refactor: restrict visibility of untested methods from internal to private and remove redundant comment in CardHolder

- remove redundant comment in draw(), already explained in onDrawFailed()

miooo0o committed:
refactor(application, game-result): extract DealerEarning init logic from GameResult

miooo0o committed:
refactor(game-manager, test): separate OutputView from playPlayerRound and remove println() in tests

miooo0o committed:
refactor(player-factory): convert interface to fun interface to enable extensibility via lambda

miooo0o committed:
docs(readme): add/update checklist for step 2-2 tasks from review

miooo0o commented:
### **Step 2-2 Review Feedback**

Hi PK, 
Thank you for helping me see where I can improve.I learn a lot from you!

#### The following changes are included in this set of commits:
- Limited to 5 commits
- Changed visibility of untested methods to `private`
- Extracted `DealerEarning` init from `GameResult`
- Removed `println()` and` OutputView` calls from tests
- Cleaned up unnecessary test
- Deleted redundant comment in `CardHolder`
- Changed BasePlayerFactory to `fun interface`

---

- This was my first time using git squash. I used to commit everything at once, so I focused only on splitting commits. I hadn't really thought about how to merge them back. It feels good to learn a new tool. 
- Extracting `DealerEarning` from `GameResult` was surprisingly simple. lovely.

As I mentioned above, if there's anything you'd like me to try out in Step 2, feel free to let me know.
I can work on it little by little over the week/end!

Thank you again and have a great day.

pkeugine reviewed (approved):
Hello Mina,
Nice work refactoring the project in 5 commits 👍

Like I said before, I think it's time we finish this blackjack mission. I hope you do well on your next missions, and if there are any questions or thought, feel free to DM me anytime.

Happy Hacking! 🌟🌟🌟
